### PR TITLE
Tests and misc updates for L2 withdrawals root

### PR DESCRIPTION
**Pull request by vdamle**
Targeted to be merged into: https://github.com/ethereum-optimism/op-geth/pull/383

for TODO work items in `op-geth` documented in: https://github.com/ethereum-optimism/optimism/issues/12044

### CODE REVIEW

File: beacon/engine/types.go [hunk 1]
```diff
@@ -275,7 +275,7 @@ func ExecutableDataToBlockNoHash(data ExecutableData, versionedHashes []common.H
 	// ExecutableData before withdrawals are enabled by marshaling
 	// Withdrawals as the json null value.
 	var withdrawalsRoot *common.Hash
-	if data.WithdrawalsRoot != nil {
+	if config.IsOptimismIsthmus(data.Timestamp) && data.WithdrawalsRoot != nil {
 		if data.Withdrawals == nil || len(data.Withdrawals) != 0 {
 			return nil, fmt.Errorf("attribute WithdrawalsRoot was set. Expecting non-nil empty withdrawals list, but got %v", data.Withdrawals)
```
Comment (vdamle) [inline beacon/engine/types.go]: There are a bunch of tests in `eth/catalyst/api_test.go` which explicitly set a withdrawals list assuming a post-shanghai block. Example: https://github.com/ethereum-optimism/op-geth/blob/optimism/eth/catalyst/api_test.go#L1114-L1131

These tests obviously fail because we require `data.Withdrawals` to be an empty list or nil. Passing in the chain config here allows us to check if we're at a Isthmus block and enforce the condition only if that's true.

**Comment by vdamle** [comment]
@protolambda In order to update the genesis handling when reading from superchain-registry: https://github.com/ethereum-optimism/op-geth/blob/optimism/core/superchain.go#L30-L46 , I will need to make a change to SCR to:

* become aware of `Isthmus` fork.
* read the `withdrawalsRoot` from the genesis data, if Isthmus is active.

Going to work on that PR, so there's a TBD in `op-geth` to update that portion of the genesis after the SCR code is in.

File: core/rlp_test.go [hunk 2]
```diff
@@ -195,3 +200,35 @@ func BenchmarkHashing(b *testing.B) {
 		b.Fatalf("hash wrong, got %x exp %x", got, exp)
 	}
 }
+
+func TestBlockRlpEncodeDecode(t *testing.T) {
```
Comment (tynes) [inline core/rlp_test.go:203]: Is there already a roundtrip RLP encoding test for blocks someplace in this codebase? What does this cover specifically that would not be covered otherwise?

File: core/types/block.go [hunk 3]
```diff
@@ -84,7 +86,7 @@ type Header struct {
 	GasLimit    uint64         `json:"gasLimit"         gencodec:"required"`
 	GasUsed     uint64         `json:"gasUsed"          gencodec:"required"`
 	Time        uint64         `json:"timestamp"        gencodec:"required"`
-	Extra       []byte         `json:"extraData"        gencodec:"required"`
+	Extra       []byte         `json:"extraData"        gencodec:"required" deep:"-"`
```
Comment (tynes) [inline core/types/block.go]: What does the `deep` tag mean here?

File: core/genesis.go [hunk 4]
```diff
@@ -477,22 +491,23 @@ func (g *Genesis) IsVerkle() bool {
...
 	var err error
 	if g.StateHash != nil {
 		if len(g.Alloc) > 0 {
 			panic(fmt.Errorf("cannot both have genesis hash %s "+
 				"and non-empty state-allocation", *g.StateHash))
 		}
-		root = *g.StateHash
-	} else if root, err = hashAlloc(&g.Alloc, g.IsVerkle()); err != nil {
+		// TODO - need to get the storage root of the L2ToL1MessagePasser contract?
+		stateRoot = *g.StateHash
```
Comment (vdamle) [inline core/genesis.go:512]: Going to need some input here @protolambda . If we find just the [StateHash](https://github.com/ethereum-optimism/superchain-registry/blob/main/superchain/superchain.go#L497-L500), is there a way to get the storage root of the message passer?

File: core/types/block_test.go [hunk 5]
```diff
@@ -423,3 +428,115 @@ func TestCheckTransactionConditional(t *testing.T) {
...
+	check("Root", decoded.Root, header.Root)
+	check("TxHash", decoded.TxHash, header.TxHash)
+	check("ReceiptHash", decoded.ReceiptHash, header.ReceiptHash)
+	check("WithdrawalsHash", decoded.WithdrawalsHash, header.WithdrawalsHash)
+	check("Bloom", decoded.Bloom, header.Bloom)
+	check("Difficulty", decoded.Difficulty, header.Difficulty)
+	check("Number", decoded.Number, header.Number)
+	check("GasLimit", decoded.GasLimit, header.GasLimit)
+	check("GasUsed", decoded.GasUsed, header.GasUsed)
+	check("Time", decoded.Time, header.Time)
```
Comment (tynes) [inline core/types/block_test.go]: These are a lot of copy/paste lines of code that could be deduplicated, also surprised there isn't this test coverage already in geth

File: core/types/transaction_conditional_test.go [hunk 6]
```diff
@@ -37,26 +37,26 @@ func TestTransactionConditionalCost(t *testing.T) {
 		{
 			name: "default cost per account",
 			cond: TransactionConditional{KnownAccounts: map[common.Address]KnownAccount{
-				common.Address{19: 1}: KnownAccount{},
-				common.Address{19: 2}: KnownAccount{},
+				{19: 1}: {},
```
Comment (tynes) [inline core/types/transaction_conditional_test.go]: This is unnecessary diff that adds noise for reviewers, its not related to the feature in this PR so we should try to reduce diff in this PR

File: beacon/types/exec_data_test.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+		Withdrawals:     []*types.Withdrawal{},
+		WithdrawalsRoot: &withdrawalsHash1,
+	}
+
+	executableDataSamples = []engine.ExecutableData{
+		isthumusExecData,
+	}
+)
+
+func TestExecutableDataJSONEncodeDecode(t *testing.T) {
```
Comment (tynes) [inline beacon/types/exec_data_test.go]: Roundtrip JSON serialization doesn't already exist in the codebase?

Comment (vdamle) [reply beacon/types/exec_data_test.go]: No, I didn't find any in the codebase. `eth/catalyst/api_test.go` has a few `ExecutableData` -> `Block` and vice versa but none for JSON serialization.

File: core/rlp_test.go (hunk 2 above)
Comment (vdamle) [reply core/rlp_test.go:203]: There is an existing RLP encoding roundtrip that is part of a test that only checks whether transactions/uncles can be iterated -- and it uses `TestChainConfig` that does not have op-stack forks set. This test aims to exercise the behavior of creating a new block when Isthmus hard fork is active: https://github.com/ethereum-optimism/op-geth/pull/399/files#diff-e177ddb5610d1f3426c15cb512999613785b20838954e48187883e6263bb282bR299-R304 and making sure that the new header field is as expected after the encode/decode round trip.

File: core/types/block.go (hunk 3 above)
Comment (vdamle) [reply core/types/block.go]: My bad, I missed including a comment here. One of the subtleties that was exposed from the round trip test I added is that an empty slice set for the `extraData` is not preserved after RLP encode/decode and the decoded field is a `nil` slice. Obviously, from a semantic point of view, it doesn't make a difference since nil slices work just fine for checking lengths etc. but the test fails any deep comparison using `reflect.deepEqual`. I ended up having to use https://pkg.go.dev/github.com/go-test/deep to check for the differing field. the `deep:"-"` tag ignores the `extraData` field when using deep compare in that test. Will add a comment as well to clarify.

File: core/types/transaction_conditional_test.go (hunk 6 above)
Comment (vdamle) [reply core/types/transaction_conditional_test.go]: I was meaning to get rid of the pesky vscode warnings about dead code, I do see your point that this is noise. Happy to revert this change from this PR and create a separate one to address this cosmetic issue.

File: core/types/block_test.go (hunk 5 above)
Comment (vdamle) [reply core/types/block_test.go]: Refactored and de-duped checks and making a test header. Unless I've completely missed it, I didn't find this round trip test in existing code.

File: core/types/block.go (hunk 3 above)
Comment (tynes) [reply core/types/block.go]: We should be careful with adding these sorts of tags to consensus critical types when they are just useful for testing

File: core/genesis.go (hunk 4 above)
Comment (protolambda) [reply core/genesis.go:512]: The `stateHash` is only there for pre-bedrock chains, which also means pre-isthmus. So no need to retrieve/set the withdrawals hash in those chains. You can update the comment to reflect that dependency. And maybe add a error or panic if someone tries to use state-hash for a genesis that is post-isthmus.

File: core/types/block.go (hunk 3 above)
Comment (protolambda) [reply core/types/block.go]: Can you maybe just json-encode both values, and compare the json? That way you don't have to introduce special tags or libraries.

File: core/genesis.go [hunk 8]
```diff
@@ -155,15 +158,24 @@ func hashAlloc(ga *types.GenesisAlloc, isVerkle bool) (common.Hash, error) {
...
 		}
 	}
-	return statedb.Commit(0, false)
+
+	stateRoot, err := statedb.Commit(0, false)
+	if err != nil {
+		return common.Hash{}, common.Hash{}, err
+	}
+	// get the storage root of the L2ToL1MessagePasser contract
+	storageRootMessagePasser := statedb.GetStorageRoot(params.OptimismL2ToL1MessagePasser)
```
Comment (protolambda) [inline core/genesis.go]: This storage lookup should be conditional on isthmus. Although it may return a zeroed hash, we don't want code-paths to run if they don't need to run, to derisk changes related to a hardfork before/outside said hardfork.

File: core/genesis.go [hunk 9]
```diff
@@ -177,17 +189,19 @@ func flushAlloc(ga *types.GenesisAlloc, triedb *triedb.Database) (common.Hash, e
...
 		}
 	}
-	root, err := statedb.Commit(0, false)
+	stateRoot, err := statedb.Commit(0, false)
 	if err != nil {
-		return common.Hash{}, err
+		return common.Hash{}, common.Hash{}, err
 	}
+	// get the storage root of the L2ToL1MessagePasser contract
+	storageRootMessagePasser := statedb.GetStorageRoot(params.OptimismL2ToL1MessagePasser)
```
Comment (protolambda) [inline core/genesis.go]: Same here, please make it conditional on the chain config

File: core/genesis_test.go [hunk 10]
```diff
@@ -223,13 +223,16 @@ func TestReadWriteGenesisAlloc(t *testing.T) {
 			{1}: {Balance: big.NewInt(1), Storage: map[common.Hash]common.Hash{{1}: {1}}},
 			{2}: {Balance: big.NewInt(2), Storage: map[common.Hash]common.Hash{{2}: {2}}},
 		}
-		hash, _ = hashAlloc(alloc, false)
+		stateRoot, storageRootMessagePasser, _ = hashAlloc(alloc, false)
 	)
+	if storageRootMessagePasser.Cmp(common.Hash{}) != 0 {
```
Comment (protolambda) [inline core/genesis_test.go]: `common.Hash` is an array and the zero value can be compared in place like so, to avoid the `Cmp`:

```suggestion
	if storageRootMessagePasser != (common.Hash{}) {
```

File: core/types/block.go [hunk 11]
```diff
@@ -294,15 +296,24 @@ func NewBlock(header *Header, body *Body, receipts []*Receipt, hasher TrieHasher
...
-		b.header.WithdrawalsHash = &hash
+	if config.IsIsthmus(header.Time) {
+		if withdrawals == nil || len(withdrawals) > 0 {
+			log.Crit("expected non-nil empty withdrawals operation list in Isthmus, but got: %v", body.Withdrawals)
+		}
+		b.header.WithdrawalsHash = header.WithdrawalsHash
 		b.withdrawals = slices.Clone(withdrawals)
+	} else {
+		// pre-Canyon
+		if withdrawals == nil {
```
Comment (protolambda) [inline core/types/block.go]: If you change this into an `} else if {` the diff will be smaller, as it avoids indenting the whole thing

File: core/types/gen_header_json.go [hunk 12]
```diff
@@ -28,7 +28,7 @@ func (h Header) MarshalJSON() ([]byte, error) {
 		GasLimit         hexutil.Uint64  `json:"gasLimit"         gencodec:"required"`
 		GasUsed          hexutil.Uint64  `json:"gasUsed"          gencodec:"required"`
 		Time             hexutil.Uint64  `json:"timestamp"        gencodec:"required"`
-		Extra            hexutil.Bytes   `json:"extraData"        gencodec:"required"`
+		Extra            hexutil.Bytes   `json:"extraData"        gencodec:"required" deep:"-"`
```
Comment (protolambda) [inline core/types/gen_header_json.go]: Can we avoid the `deep` tags?

File: core/genesis.go (hunk 9 above)
Comment (vdamle) [reply core/genesis.go]: Done.

File: core/types/gen_header_json.go (hunk 12 above)
Comment (vdamle) [reply core/types/gen_header_json.go]: Agreed, I've removed this and am now checking a few individual fields in the header.

File: core/genesis.go (hunk 8 above)
Comment (vdamle) [reply core/genesis.go]: Done.

File: core/types/block.go (hunk 3 above)
Comment (vdamle) [reply core/types/block.go]: Agreed, I've removed this and am now checking a few individual fields in the header.

File: core/genesis.go [hunk 13]
```diff
@@ -549,8 +579,17 @@ func (g *Genesis) toBlockWithRoot(root common.Hash) *types.Block {
 			head.RequestsHash = &types.EmptyRequestsHash
 			requests = make(types.Requests, 0)
 		}
+		// If Isthmus is active at genesis, set the WithdrawalRoot to the storage root of the L2ToL1MessagePasser contract.
+		if g.IsIsthmus() {
+			if storageRootMessagePasser == (common.Hash{}) {
+				// if there was no MessagePasser contract storage, set the WithdrawalsHash to the empty hash
```
Comment (vdamle) [inline core/genesis.go:580]: The specs [note](https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/isthmus/exec-engine.md#backwards-compatibility-considerations) that `After Isthmus activation, if there was never any withdrawal contract storage, a MPT root of an empty list can be set as the withdrawalsRoot`.

**Comment by vdamle** [comment]
This [comment](https://github.com/ethereum-optimism/op-geth/pull/399#issuecomment-2438761789) is not relevant.

Because `hashAlloc()` reads the storage root of the contract, no change is needed in `superchain-registry. So the above comment can be ignored.

File: core/types/block_test.go [hunk 14]
```diff
@@ -423,3 +428,83 @@ func TestCheckTransactionConditional(t *testing.T) {
...
+func TestHeaderRLPEncoding(t *testing.T) {
+	header := getTestHeader()
+
+	encoded, err := rlp.EncodeToBytes(header)
+	if err != nil {
+		t.Fatalf("failed to encode header: %v", err)
+	}
+
+	var decoded Header
+	if err := rlp.DecodeBytes(encoded, &decoded); err != nil {
```
Comment (tynes) [inline core/types/block_test.go]: There is already a test for roundtrip encode/decode of the header here: https://github.com/ethereum/go-ethereum/blob/aa636921291744af082cefbe4dd811fb58fbde63/core/types/rlp_fuzzer_test.go#L115

I don't think anything additional is covered here with this new test

File: core/genesis.go [hunk 15]
```diff
@@ -475,24 +495,34 @@ func (g *Genesis) IsVerkle() bool {
 	return g.Config.IsVerkle(new(big.Int).SetUint64(g.Number), g.Timestamp)
 }
 
+// IsIsthmus indicates whether Isthmus is active at genesis time.
+func (g *Genesis) IsIsthmus() bool {
+	return g.Config.IsIsthmus(g.Timestamp)
+}
+
```
Comment (protolambda) [inline core/genesis.go]: IMHO we should not add additional public-API to the `Genesis` struct. It's a one-liner anyway, not worth the diff.

File: core/genesis.go [hunk 16]
```diff
@@ -475,24 +495,34 @@ func (g *Genesis) IsVerkle() bool {
...
 		if len(g.Alloc) > 0 {
 			panic(fmt.Errorf("cannot both have genesis hash %s "+
 				"and non-empty state-allocation", *g.StateHash))
 		}
-		root = *g.StateHash
-	} else if root, err = hashAlloc(&g.Alloc, g.IsVerkle()); err != nil {
+		// stateHash is only relevant for pre-bedrock (and hence pre-isthmus) chains.
+		// we bail here since this is not a valid usage of StateHash
+		if g.IsIsthmus() {
+			panic(fmt.Errorf("stateHash usage disallowed in isthmus chain"))
```
Comment (protolambda) [inline core/genesis.go]:
```suggestion
			panic(fmt.Errorf("stateHash usage disallowed in chain with isthmus active at genesis"))
```

File: core/types/block.go [hunk 17]
```diff
@@ -294,7 +296,14 @@ func NewBlock(header *Header, body *Body, receipts []*Receipt, hasher TrieHasher
 		}
 	}
 
-	if withdrawals == nil {
+	if config.IsIsthmus(header.Time) {
+		if withdrawals == nil || len(withdrawals) > 0 {
+			log.Crit("expected non-nil empty withdrawals operation list in Isthmus, but got: %v", body.Withdrawals)
+		}
```
Comment (protolambda) [inline core/types/block.go]: A `panic` is better than a `log.Crit`, since `log.Crit` has program-exit behavior, which is bad with testing and unexpected user input.

File: core/types/block.go [hunk 18]
```diff
@@ -294,7 +296,14 @@ func NewBlock(header *Header, body *Body, receipts []*Receipt, hasher TrieHasher
...
+		b.header.WithdrawalsHash = header.WithdrawalsHash
+		b.withdrawals = slices.Clone(withdrawals)
```
Comment (protolambda) [inline core/types/block.go]: Since we already assert it's non-nil and empty, doing a `make(types.Withdrawals, 0)` may be more readable than a `slices.Clone`.

File: core/types/block_test.go [hunk 19]
```diff
@@ -423,3 +428,83 @@ func TestCheckTransactionConditional(t *testing.T) {
...
+		t.Fatalf("failed to decode header: %v", err)
+	}
+
+	checkFields(t, &decoded, header)
+}
+
+func TestHeaderJSONEncoding(t *testing.T) {
```
Comment (tynes) [inline core/types/block_test.go]: I didn't see this covered already when looking around the source code, I wish we had better test coverage. It could make sense to turn this into a fuzz test rather than hardcoding the test header, similar to https://github.com/ethereum/go-ethereum/blob/aa636921291744af082cefbe4dd811fb58fbde63/core/types/deposit_test.go#L40

File: core/rlp_test.go [hunk 20]
```diff
@@ -195,3 +200,42 @@ func BenchmarkHashing(b *testing.B) {
...
+	zeroTime := uint64(0)
+
```
Comment (tynes) [inline core/rlp_test.go:205]: How does this compare to what is in https://github.com/ethereum/go-ethereum/blob/aa636921291744af082cefbe4dd811fb58fbde63/core/types/rlp_fuzzer_test.go#L118?

File: core/rlp_test.go [hunk 21]
```diff
@@ -195,3 +200,42 @@ func BenchmarkHashing(b *testing.B) {
...
+			t.Errorf("%s mismatch", f)
+			t.Errorf("Got: %+v", got)
+			t.Errorf("Want: %+v", want)
+		}
+	}
+
+	// There's an odd inconsistency in the way `ExtraData` field, when it is empty is returned after
+	// rlp encode-decode roundtrip. The input is an empty byte slice, but the output is a nil slice,
+	// due to which the reflect.DeepEqual fails. So, we compare a few fields in the header/block manually.
+	// for triaging this, "https://pkg.go.dev/github.com/go-test/deep" was useful since it spits out the
```
Comment (tynes) [inline core/rlp_test.go]: Does it make sense to hash and compare the hash since it commits to all fields?

File: core/genesis.go [hunk 22]
```diff
@@ -549,8 +579,17 @@ func (g *Genesis) toBlockWithRoot(root common.Hash) *types.Block {
 			head.RequestsHash = &types.EmptyRequestsHash
 			requests = make(types.Requests, 0)
 		}
+		// If Isthmus is active at genesis, set the WithdrawalRoot to the storage root of the L2ToL1MessagePasser contract.
+		if g.IsIsthmus() {
+			if storageRootMessagePasser == (common.Hash{}) {
```
Comment (tynes) [inline core/genesis.go:579]: Will it be `common.Hash{}` or [types.EmptyRootHash](https://github.com/ethereum/go-ethereum/blob/aa636921291744af082cefbe4dd811fb58fbde63/core/types/hashes.go#L27) at this abstraction?

File: beacon/types/exec_data_test.go (hunk 7 above)
Comment (tynes) [inline beacon/types/exec_data_test.go]: This could be good as a fuzz test

File: core/types/block_test.go (hunk 14 above)
Comment (vdamle) [reply core/types/block_test.go]: You are right, I don't believe I saw this when I was checking to see if a test exists. I'll remove the rlp roundtrip test.

File: core/rlp_test.go (hunk 20 above)
Comment (vdamle) [reply core/rlp_test.go:205]: I'm specifically testing the correctness of rlp encode/decode of a block generated after isthmus hard fork is active. I'm guessing the fuzzer test would probably cover this as well? (all possible values for the withdrawals and withdrawalsHash fields?)

File: beacon/types/exec_data_test.go (hunk 7 above)
Comment (vdamle) [reply beacon/types/exec_data_test.go]: As discussed with @tynes , will move this to a fuzz test in a follow up.

File: core/genesis.go (hunk 22 above)
Comment (vdamle) [reply core/genesis.go:579]: My understanding is that if there is no contract deployment, this woulld be an empty hash and not a empty root hash?

File: beacon/engine/types.go [hunk 23]
```diff
@@ -275,7 +275,7 @@ func ExecutableDataToBlockNoHash(data ExecutableData, versionedHashes []common.H
 	// ExecutableData before withdrawals are enabled by marshaling
 	// Withdrawals as the json null value.
 	var withdrawalsRoot *common.Hash
-	if data.WithdrawalsRoot != nil {
+	if config.IsOptimismIsthmus(data.Timestamp) && data.WithdrawalsRoot != nil {
```
Comment (tynes) [inline beacon/engine/types.go]: I am surprised that this doesn't break any tests as when op stack is disabled, this branch will never be hit

File: core/rlp_test.go [hunk 24]
```diff
@@ -195,3 +200,36 @@ func BenchmarkHashing(b *testing.B) {
...
+
+	check := func(f string, got, want interface{}) {
+		if equal := reflect.DeepEqual(got, want); equal != true {
+			t.Errorf("%s mismatch", f)
+			t.Errorf("Got: %+v", got)
+			t.Errorf("Want: %+v", want)
+		}
+	}
+
+	// There's an odd inconsistency in the way `ExtraData` field, when it is empty is returned after
```
Comment (tynes) [inline core/rlp_test.go]: Is this a dead comment?

File: core/types/gen_transaction_conditional_json.go [hunk 25]
```diff
@@ -39,7 +40,11 @@ func (t *TransactionConditional) UnmarshalJSON(input []byte) error {
 		TimestampMax   *math.HexOrDecimal64  `json:"timestampMax,omitempty"`
 	}
 	var dec TransactionConditional
-	if err := json.Unmarshal(input, &dec); err != nil {
+	// --- Not Generated. Disallow unknown fields
```
Comment (tynes) [inline core/types/gen_transaction_conditional_json.go]: Why is this diff included? Its not related to this feature and also we should not be manually modifying autogenerated files

File: core/rlp_test.go [hunk 26]
```diff
@@ -195,3 +200,36 @@ func BenchmarkHashing(b *testing.B) {
...
+	// rlp encode-decode roundtrip. The input is an empty byte slice, but the output is a nil slice,
+	// due to which the reflect.DeepEqual fails. So, we compare the hash instead of each individual field.
+	// for triaging this, "https://pkg.go.dev/github.com/go-test/deep" was useful since it spits out the
+	// exact field that is different.
+	check("block hash", decoded.Hash(), block.Hash())
```
Comment (tynes) [inline core/rlp_test.go]: `assert.Equal` rather than keeping around a function only used once?

File: core/rlp_test.go (hunk 24 above)
Comment (vdamle) [reply core/rlp_test.go]: I thought it's a valuable pointer for anyone touching this test in the future, because if there's an attempt made to compare individual fields like I did, they'll step into the same issue. Happy to remove it though.

File: core/types/gen_transaction_conditional_json.go (hunk 25 above)
Comment (vdamle) [reply core/types/gen_transaction_conditional_json.go]: Seems to be related to the go version on my Mac. It's not intended, will remove this.

File: beacon/types/exec_data_test.go (hunk 7 above)
Comment (vdamle) [reply beacon/types/exec_data_test.go]: Done - Moved `beacon.Header` and `types.xyz` encode/decode rountrip to a fuzz test.

File: core/genesis.go (hunk 22 above)
Comment (protolambda) [reply core/genesis.go:579]: +1, if the contract is absent then I believe the storage-root is zeroed.

File: beacon/engine/types.go (hunk 23 above)
Comment (protolambda) [reply beacon/engine/types.go]: The provided withdrawals-root is fine to ignore when the same root is re-computed from the actual withdrawals.
That said, yes this `IsOptimismIsthmus(data.Timestamp) &&` shouldn't be here. Instead, I think we should have a separate `if IsOptimismIsthmus`, that then checks if the withdrawals-root is set, and error otherwise. And leave the existing withdrawals-root check the same, since it just prevents recomputing it if it's set already.

**Review by protolambda — COMMENTED** [review]
One change to make, then good to merge into the base branch. Will do another review there, and then with the monorepo side ready (don't want to block updates to latest op-geth due to breaking changes in block-construction signatures) we can merge it `optimism` branch.

File: beacon/engine/types.go (hunk 23 above)
Comment (vdamle) [reply beacon/engine/types.go]: @protolambda Your PR [requires](https://github.com/ethereum-optimism/op-geth/pull/383/files#diff-50e0dca0bd0ac667ca6a4a3b762c9346030f346fdd9b48f3f83be675a43423eaR278-R283) that the withdrawals list be empty and returns an error otherwise.

To @tynes 's point, as-is, that breaks a bunch of existing tests [here](https://github.com/ethereum-optimism/op-geth/blob/optimism/eth/catalyst/api_test.go#L1341-L1350) that explicitly set a non-zero withdrawals list.

I was thinking that the check you added in your PR should only be enforced if Isthmus is active (which obviously implies that op-stack is enabled). Am I missing something?

Comment (protolambda) [reply beacon/engine/types.go]: That's an error in my PR draft, we should change that. You're right that there needs to be an optimism fork check on it. But the one like implemented here is broken as described above.

File: beacon/engine/types.go [hunk 27]
```diff
@@ -275,15 +275,15 @@ func ExecutableDataToBlockNoHash(data ExecutableData, versionedHashes []common.H
...
-	} else if data.Withdrawals != nil {
+	if config.IsOptimismIsthmus(data.Timestamp) && data.WithdrawalsRoot == nil {
+		return nil, fmt.Errorf("attribute WithdrawalsRoot is required for Isthmus blocks")
+	}
+	if data.Withdrawals != nil {
 		h := types.DeriveSha(types.Withdrawals(data.Withdrawals), trie.NewStackTrie(nil))
 		withdrawalsRoot = &h
+	} else if data.WithdrawalsRoot != nil {
+		h := *data.WithdrawalsRoot // copy, avoid any sharing of memory
+		withdrawalsRoot = &h
```
Comment (protolambda) [inline beacon/engine/types.go]: This still doesn't work, since an empty withdrawals list here would result in the empty-root, even if the withdrawals root is set. We need to prioritize the withdrawals-root, if set, like it used to be before. And only compute the withdrawals root manually in the else case.

**Review by protolambda — APPROVED** [review]

File: core/rlp_test.go (hunk 20 above)
Comment (protolambda) [reply core/rlp_test.go:205]: I think we can move forward with merging this in the underlying branch, so we have one target of testing / rebasing. Marking this as resolved, but will give all the tests another review once into the base PR.
